### DISCUSSION TITLE
serialbattery stop charging at 99.1%

### DISCUSSION THREAD

**Discussion by hansgeorgk** [Louisvdw/dbus-serialbattery#126] [Q&A] [answered]
Hi,

very nice work - very much appreciated :-)
Cool stuff to have the charge and discharge limits changing depending on the SOC - took me a while to realize why the numbers are different to my settings ;-)
I am using Venus OS (2.86), with Easysolar (Multiplus and MPPT), BMV and DalyBMS with EVE Cells.
DVCC is set to ON. All devices are on latest Firmware releases.

I realized that the serialbattery stops charging at 99.1%. The setting in IO goes to Allow charging "No"
<img width="517" alt="Bildschirmfoto 2022-06-02 um 17 55 35" src="https://user-images.githubusercontent.com/100286224/171673957-39202ba9-2823-4268-a6ee-dcd2a3a999c7.png">

<img width="304" alt="Bildschirmfoto 2022-06-01 um 16 54 00" src="https://user-images.githubusercontent.com/100286224/171676026-7deb3b90-c80c-4845-ad9d-72f9143de41d.png">

I have set the limits in utils.py, but it seems like they are not reached

```
# Constants - Need to dynamically get them in future
DRIVER_VERSION = 0.10
DRIVER_SUBVERSION = ''
zero_char = chr(48)
degree_sign = u'\N{DEGREE SIGN}'
# Cell min/max voltages - used with the cell count to get the min/max battery voltage
MIN_CELL_VOLTAGE = 2.5
MAX_CELL_VOLTAGE = 3.65
# battery Current limits
MAX_BATTERY_CURRENT = 100.0
MAX_BATTERY_DISCHARGE_CURRENT = 200.0
# Invert Battery Current. Default non-inverted. Set to -1 to invert
INVERT_CURRENT_MEASUREMENT = -1
```

Max Cell voltage goes only up to 3.356 and the available capacity stays below the installed (designed) capacity. By the way, the installed is not shown ;-(

thanks a lot
cheers

**Comment by Louisvdw** [comment, marked as answer]
Hi Hans-Georg
When your battery shows the Allow charging "No" it means that the settings in your BMS is so that something triggered the BMS to disconnect the charging. 

First thing would be to check your Alarm Logs in VRM which should show if any alarms were triggered. That will give you an isea what happened.

I think your issue is that there is a BMS setting is not correct. Your BMS hardware settings might trigger a stop to charging if your cells a different voltage that what you set the driver to do. You should keep your driver to smaller limits to prevent your BMS from triggering protections. You will have a much happier battery that way.

Normally I would not suggest you use 3.65V for the cell max voltage, but from you min/max graph that does not seem to be your issue here. I would however suggest that you revert that to the default 3.45V setting and see if that helps. See [the FAQ](https://github.com/Louisvdw/dbus-serialbattery/wiki/FAQ#why-is-the-max-cell-voltage-set-to-345v) for more details if you want.

You did not mention which BMS you have. They all have small differences and not all have the capacity available to read.

**Comment by fabian-lauer** [comment]
Hi @Louisvdw,
hi @hansgeorgk,

first thanks for the cool module, really great.

But I am facing almost the same problem as @hansgeorgk: loading stucks at 99.1%, "Allow charge: NO".
In addition the MultiPlus 2 is always in mode "Bulk" - based on VE.Configure it should switch after 95% to "Absorption".

My config:
- 16 LiFePo4 cells
- Daly Smart BMS connected via RS486 (USB-Adapter)
- Multiplus 2
- utils.py
```
# Constants - Need to dynamically get them in future
DRIVER_VERSION = 0.10
DRIVER_SUBVERSION = ''
zero_char = chr(48)
degree_sign = u'\N{DEGREE SIGN}'
# Cell min/max voltages - used with the cell count to get the min/max battery voltage
MIN_CELL_VOLTAGE = 3.1
MAX_CELL_VOLTAGE = 3.45
# battery Current limits
MAX_BATTERY_CURRENT = 180.0
MAX_BATTERY_DISCHARGE_CURRENT = 180.0
# Invert Battery Current. Default non-inverted. Set to -1 to invert
INVERT_CURRENT_MEASUREMENT = -1
```

Pics of the current config in Daly BMS: 
![IMG_6623](https://user-images.githubusercontent.com/101807694/172135330-9ebc780d-2524-46b1-9441-d2d63378c782.jpg)
![IMG_6624](https://user-images.githubusercontent.com/101807694/172135334-cc9e7e9a-1188-435a-b3d2-a0705ae54b3b.jpg)
![IMG_6625](https://user-images.githubusercontent.com/101807694/172135340-bf8726db-6bee-437a-8cf4-174d8685b208.jpg)
![IMG_6626](https://user-images.githubusercontent.com/101807694/172135343-cd607858-9682-4341-bebe-6e142f8739b6.jpg)
![IMG_6627](https://user-images.githubusercontent.com/101807694/172135346-b49cd048-33dc-4a58-8b0b-a0fb247551b5.jpg)
![IMG_6628](https://user-images.githubusercontent.com/101807694/172135349-63e391f8-9b23-48de-95b1-4a8f3d4bf3c6.jpg)

Would be happy, if someone has an idea, because I have the feeling I configured something wrong :-/

Thx.

**Reply by Louisvdw** [reply]
Hi @fabian-lauer 
Great sending all the detail in all the BMS settings. 
Can you also check that inside your VRM if there are any Alarms in the Alarm Logs. Post those here as well.
Also in VRM under Advanced, enable the Battery Monitor widgets for Battery SOC, BMS Charge and Discharge Limits, BMS Min/Max Cell Voltage, Battery Voltage and Current and upload those as well.

Regarding the question of Bulk and Absorption those are Lead Acid battery things. As you have lithium batteries these does not apply. The BMS manage the battery cells. Lead Acids does not have an BMS and required the charger to have some logic to manage the battery in some way. This was where Bulk, Float and Absorption voltages came in. 
If you set these in your Multiplus they are the default that the system will revert to, but any assistant that you load, or the ESS and in our case DVCC override those.

**Reply by fabian-lauer** [reply]
Hi - thanks for the little explanation regarding the status of bulk, etc. Got it.

Attached you find the pictures - have to say that I tried to set SOC in Daly BMS app to 100 what worked out - battery had to discharge a bit (because of loads) and is now still charging but got stuck at 99.6% - so same thing - different number

<img width="245" alt="bat_limits" src="https://user-images.githubusercontent.com/101807694/172141724-514b3072-9bdf-4835-acba-39b5d086c528.PNG">
<img width="247" alt="bat_min-max" src="https://user-images.githubusercontent.com/101807694/172141725-e8c224eb-3f92-4863-8a3a-777211960637.PNG">
<img width="247" alt="bat_soc" src="https://user-images.githubusercontent.com/101807694/172141729-e056d0c3-a9de-48e7-97cf-9efa232817e7.PNG">
<img width="245" alt="bat_state" src="https://user-images.githubusercontent.com/101807694/172141731-90affa88-e2cb-4805-9ef6-666919fc8695.PNG">
<img width="248" alt="bat_volt-current" src="https://user-images.githubusercontent.com/101807694/172141732-fae0a2e0-8b5b-4222-8c8f-7ebda5bf10fc.PNG">

**Reply by fabian-lauer** [reply]
@Louisvdw , I guess your request "Charge + Discharge Limits" triggered my brain a bit :-) 

I just found why "Allow Charge" is set to no - its in the code - if SOC > 99 it is set to "no":
https://github.com/Louisvdw/dbus-serialbattery/blob/63a450baa532089191da213c94b55c9250cfe96b/etc/dbus-serialbattery/battery.py#L116

Or am I wrong?

**Reply by Louisvdw** [reply]
You sometime see this on production batteries as well. They keep the SOC just below 100%. I would not worry too much. It might be a rounding error in the calculations.

I do see that your Balaner is turned Off. Do you have an external balancer on this battery? The Daly's balancer is not one of the best, but you do need something to keep the cells level.

The min/max cell voltage does show that die max was at 3.45V while the min was still around 3.35V so some balancing would be good for the cells. But overall it seems fine.

You could lower your max cell voltage. This gives your system more time to reach 100% SOC with less stress.
For instance my second battery bank that I added to test the multi bank feature include some old cells that don't all have the same capacity. By reducing my cell voltage I aim for I can get the battery to 100% SOC without some cells reaching overvoltage.
MAX_CELL_VOLTAGE = 3.40

**Reply by Louisvdw** [reply]
@fabian-lauer that is true. The flag does get set there. Perhaps what happens is that the Daly's value comes through as with decimals while the other BMS I tested does not. I'll investigate that a bit to see if I can get an answer.
If there are no decimal value, then it should only activate the flag when 100% is reached which is what I see.

**Reply by fabian-lauer** [reply]
@Louisvdw thanks for the hint with the lower-cell voltage - will try.
Yes also have an external balancer installed - BMS balancer is not disabled in general, just for a certain threshold (0.030v).

**Reply by fabian-lauer** [reply]
@Louisvdw it looks like that the SOC is stored with a decimal - at least in DBUS (checked with dbus-spy on ssh):
![image](https://user-images.githubusercontent.com/101807694/172145955-a58a7669-424d-4d2f-bbfe-eac4573d1bb3.png)
